comfy/graph_utils.py

```diff
@@ -0,0 +1,140 @@
+def is_link(obj):
+    if not isinstance(obj, list):
+        return False
+    if len(obj) != 2:
+        return False
+    if not isinstance(obj[0], str):
+        return False
+    if not isinstance(obj[1], int) and not isinstance(obj[1], float):
+        return False
+    return True
+
+# The GraphBuilder is just a utility class that outputs graphs in the form expected by the ComfyUI back-end
+class GraphBuilder:
+    _default_prefix_root = ""
+    _default_prefix_call_index = 0
+    _default_prefix_graph_index = 0
```

Comment:
> The expansion function needs to support actual expansion, returning literal values, and any mix between the two. At that point, it supports everything the normal evaluation function supports anyway.

Hm, the whole idea of separating expansion and execution is that the expansion function doesn't return any literal values, otherwise there is no actual separation. I guess you are referring to cases where a node needs expansion to compute a subset of its outputs, but the rest can be ready immediately? If so, it's a weird case to support. To me it indicates that we're combining the functionality of two separate nodes into one, and can solve it by refactoring. WDYT? Also, even now this case doesn't seem to be supported (judging by this [`elif`]( https://github.com/guill/ComfyUI/blob/1f065889053a9b1328dde614a729eae886ee60a3/execution.py#L217)), so making expansion a separate function only makes it more explicit.

> When this expansion function exists, the normal evaluation function wouldn't be called. The two are mutually exclusive.

I think that's absolutely fine if it's communicated to the authors of custom nodes.

Reply:
> I guess you are referring to cases where a node needs expansion to compute a subset of its outputs, but the rest can be ready immediately? If so, it's a weird case to support. To me it indicates that we're combining the functionality of two separate nodes into one, and can solve it by refactoring. WDYT? 

The issue is that if we don't inherently support returning literals -- if every return value has to be the result of a subgraph evaluation -- implementations that would be otherwise simple (like dynamically adding a number of LoRAs and returning a string containing the LoRAs you used) would require the addition of new "literal" nodes that do nothing but return the passed literal value.

The issue you found with explicit return values being dropped only applies when mixing return formats -- using the dictionary syntax for returning one value, but the tuple syntax when returning another. All of the other cases (either returning results from a subgraph on one output and literal results on another or mixing node expansions and literal returns across a list) already function correctly. I've added a test that demonstrates that (and fixed the issue you found).

I'm not totally opposed to requiring that node authors use a different method name for node expansion, but given that it still has to support literal return values (and I feel very strongly that it does), I think it'll make the code more difficult to read rather than easier. The one benefit of it I see is the addition of the factory argument. I'm just not sure if that benefit justifies the added complexity.

Reply:
I looked through your examples and https://github.com/BadCafeCode/execution-inversion-demo-comfyui/blob/e3dccd4616467985801498979e326b9f88f1857e/nodes.py#L65-L90 convinced me that it's a good idea. While it could be refactored to support no-immediate-value expansion, the refactored version would be more complicated. This means that we have to treat link outputs as first-class outputs, and the entire thing falls into place. Let's consider this resolved  👍